File: Leetcode Challenge/08_August_2020/C++/Week 3/4_Numbers With Same Consecutive Differences.cpp
//Return all non-negative integers of length N such that the absolute difference between every two consecutive digits is K.
//
//Note that every number in the answer must not have leading zeros except for the number 0 itself. For example, 01 has one leading zero and is invalid, but 0 is valid.
//
//You may return the answer in any order.
//
// 
//
//Example 1:
//
//Input: N = 3, K = 7
//Output: [181,292,707,818,929]
//Explanation: Note that 070 is not a valid number, because it has leading zeroes.
//Example 2:
//
//Input: N = 2, K = 1
//Output: [10,12,21,23,32,34,43,45,54,56,65,67,76,78,87,89,98]
// 
//
//Note:
//
//1 <= N <= 9
//0 <= K <= 9

class Solution {
public:
    vector<int> numsSameConsecDiff(int N, int K) {
        if(N == 1) {
            vector<int> fast;
            for(int i = 0; i < 10; i++) {
                fast.push_back(i);
            }
            return fast;
        }
        vector<int> comb;
        if(K != 0) {
            for(int i = 0; i < pow(2, N - 1); i++) {
                comb.push_back(i);
            }
        } else {
            comb.push_back(0);
        }
        vector<int> ans;
        bool good = true;
        for(int i = 1; i < 10; i++) {
            vector<int> n(N, 0);
            n[0] = i;
            for(int c : comb) {
                good = true;
                int m = i * pow(10, N - 1);
                for(int t = 1; t < N; t++) {
                    if((c & 1) == 0) {
                        if(n[t - 1] - K < 0) {
                            good = false;
                            break;
                        } else {
                            n[t] = n[t - 1] - K;
                            m += (n[t - 1] - K) * pow(10, N - 1 - t);
                        }
                    } else {
                        if(n[t - 1] + K > 9) {
                            good = false;
                            break;
                        } else {
                            n[t] = n[t - 1] + K;
                            m += (n[t - 1] + K) * pow(10, N - 1 - t);
                        }
                    }
                    c >>= 1;
                }
                if(good) {
                    ans.push_back(m);
                }
            }
        }
        return ans;
    }
};

File: Leetcode Challenge/08_August_2020/Rust/Week 3/4_Numbers With Same Consecutive Differences.rs
//Return all non-negative integers of length N such that the absolute difference between every two consecutive digits is K.
//
//Note that every number in the answer must not have leading zeros except for the number 0 itself. For example, 01 has one leading zero and is invalid, but 0 is valid.
//
//You may return the answer in any order.
//
// 
//
//Example 1:
//
//Input: N = 3, K = 7
//Output: [181,292,707,818,929]
//Explanation: Note that 070 is not a valid number, because it has leading zeroes.
//Example 2:
//
//Input: N = 2, K = 1
//Output: [10,12,21,23,32,34,43,45,54,56,65,67,76,78,87,89,98]
// 
//
//Note:
//
//1 <= N <= 9
//0 <= K <= 9

impl Solution {
    pub fn nums_same_consec_diff(n: i32, k: i32) -> Vec<i32> {
        if n == 1 {
            let mut fast: Vec<i32> = vec![0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
            fast
        } else{
            let mut comb: Vec<i32> = Vec::new();
            if k != 0 {
                for i in 0..2_i32.pow(n as u32 - 1) {
                    comb.push(i);
                }
            } else {
                comb.push(0);
            }
            let mut ans: Vec<i32> = Vec::new();
            let mut good: bool = true;
            for i in 1..10 {
                let mut N: Vec<i32> = vec![0; n as usize];
                N[0] = i as i32;
                for d in &comb {
                    let mut c = *d;
                    good = true;
                    let mut m: i32 = i * 10_i32.pow(n as u32 - 1);
                    for t in 1..n {
                        if (c & 1) == 0 {
                            if N[t as usize - 1] - k < 0 {
                                good = false;
                                break;
                            } else {
                                N[t as usize] = N[t as usize - 1] - k;
                                m += (N[t as usize - 1] - k) * 10_i32.pow((n - 1 - t) as u32);
                            }
                        } else {
                            if N[t as usize - 1] + k > 9 {
                                good = false;
                                break;
                            } else {
                                N[t as usize] = N[t as usize - 1] + k;
                                m += (N[t as usize - 1] + k) * 10_i32.pow((n - 1 - t) as u32);
                            }
                        }
                        c >>= 1;
                    }
                    if(good) {
                        ans.push(m);
                    } else {
                        continue;
                    }
                }
            }
            ans
        }
    }
}